### PR TITLE
A base implementation of siginfo feature

### PR DESCRIPTION
**Pull request by MichalBlk**
A simple base implementation of siginfo feature.

### CODE REVIEW

**Comment by cahirwpz** [comment]
@j-piecuch Please review, especially parts using locks.

**Review by j-piecuch — CHANGES_REQUESTED** [review]
We should sort out the locking issues before merging this.

File: sys/kern/signal.c [hunk 1]
```diff
@@ -173,17 +176,137 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
...
+  TAILQ_FOREACH (ksi, &sp->sp_info, ksi_list) {
+    TAILQ_REMOVE(&sp->sp_info, ksi, ksi_list);
+    assert(ksi->ksi_flags & KSI_FROMPOOL);
+    assert(ksi->ksi_flags & KSI_QUEUED);
+    ksiginfo_free(ksi);
+  }
+}
+
+static void sigpend_get(sigpend_t *sp, signo_t sig, ksiginfo_t *out) {
+  __sigdelset(&sp->sp_set, sig);
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/kern/signal.c:216]: Please check the case where there already is a pending 'ksigonfo' for a signal and we add a new (non empty) one with the 'sigpend_put'. We will simply overwrite the previous one.

File: sys/kern/signal.c [hunk 2]
```diff
@@ -173,17 +176,137 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
...
+
+  ksiginfo_t *kp;
+  TAILQ_FOREACH (kp, &sp->sp_info, ksi_list) {
+    if (kp->ksi_signo == sig)
+      break;
+  }
+
+  if (kp) {
+    ksi->ksi_info = kp->ksi_info;
+    ksi->ksi_flags = kp->ksi_flags | KSI_QUEUED;
```
Comment (j-piecuch) [inline sys/kern/signal.c]: Is setting the flag needed here? I think it should always be already set.

File: sys/kern/signal.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (j-piecuch) [reply sys/kern/signal.c:216]: Ok, I see it now.

File: sys/kern/signal.c [hunk 3]
```diff
@@ -173,17 +176,137 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
...
+
+  ksiginfo_t *kp;
+  TAILQ_FOREACH (kp, &sp->sp_info, ksi_list) {
+    if (kp->ksi_signo == sig)
+      break;
+  }
+
+  if (kp) {
+    kp->ksi_info = ksi->ksi_info;
+    kp->ksi_flags = ksi->ksi_flags | KSI_QUEUED;
```
Comment (j-piecuch) [inline sys/kern/signal.c:269]: You're leaking `ksi` here, you should `ksiginfo_free` it.

**Review by j-piecuch — CHANGES_REQUESTED** [review]

File: sys/kern/signal.c (hunk 3 above)
Comment (MichalBlk) [reply sys/kern/signal.c:269]: Thanks, I missed it.

**Comment by MichalBlk** [comment]
Fixed.

File: sys/kern/signal.c [hunk 4]
```diff
@@ -95,7 +98,7 @@ int do_sigaction(signo_t sig, const sigaction_t *act, sigaction_t *oldact) {
       memcpy(&p->p_sigactions[sig], act, sizeof(sigaction_t));
     /* If ignoring a pending signal, discard it. */
     if (sig_ignored(p->p_sigactions, sig))
-      __sigdelset(&td->td_sigpend, sig);
+      __sigdelset(&td->td_sigpend.sp_set, sig);
```
Comment (j-piecuch) [inline sys/kern/signal.c:101]: Shouldn't we also remove the queued `ksiginfo` here? If we don't and then send a `KSI_RAW` signal, the target process will receive a signal that should have been discarded.

**Review by j-piecuch — CHANGES_REQUESTED** [review]
A couple of issues, otherwise LGTM

File: sys/kern/signal.c [hunk 5]
```diff
@@ -221,17 +343,19 @@ void sig_kill(proc_t *p, signo_t sig) {
...
+  ksiginfo_t *kp = ksiginfo_copy(ksi);
 
   if (sig == SIGCONT) {
-    __sigdelset(&td->td_sigpend, SIGSTOP);
+    sigpend_get(&td->td_sigpend, SIGSTOP, NULL);
 
     /* In case of SIGCONT, make it pending only if the process catches it. */
     if (handler != SIG_IGN && handler != SIG_DFL)
-      __sigaddset(&td->td_sigpend, SIGCONT);
+      sigpend_put(&td->td_sigpend, kp);
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/kern/signal.c:357]: Fixed.

File: bin/utest/sigaction.c [hunk 6]
```diff
@@ -45,3 +53,47 @@ int test_sigaction_handler_returns(void) {
 
   return 0;
 }
+
+#define NCHLD 42
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply bin/utest/sigaction.c]: Fixed.

File: sys/kern/signal.c (hunk 4 above)
Comment (MichalBlk) [reply sys/kern/signal.c]: Fixed.

**Review by j-piecuch — APPROVED** [review]
LGTM

File: bin/utest/sigaction.c [hunk 7]
```diff
@@ -28,15 +32,19 @@ int test_sigaction_with_setjmp(void) {
 
 static volatile int sigusr1_handled;
 
-static void sigusr1_handler(int signo) {
+static void sigusr1_handler(int signo, siginfo_t *si, void *uctx) {
+  assert(si->si_signo == signo);
+  assert(si->si_code == SI_NOINFO);
+  assert(uctx != NULL);
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply bin/utest/sigaction.c:38]: Shouldn't SIGUSR1 info contain origin of signal, i.e. `si_pid` and `si_uid` as stated in [sigaction(2)](https://man7.org/linux/man-pages/man2/sigaction.2.html).

File: bin/utest/sigaction.c [hunk 8]
```diff
@@ -45,3 +53,47 @@ int test_sigaction_handler_returns(void) {
 
   return 0;
 }
+
+#define NCHLD 7
+
+static volatile int nreaped;
+
+static void sigchld_handler(int signo, siginfo_t *si, void *uctx) {
+  assert(si->si_code == CLD_EXITED);
```
Comment (cahirwpz) [inline bin/utest/sigaction.c:62]: This is not the only possibility. What if the child was _continued_ or _stopped_ or _killed_? Could you as well check `si_pid` and `si_status`.

File: bin/utest/sigaction.c [hunk 9]
```diff
@@ -45,3 +53,47 @@ int test_sigaction_handler_returns(void) {
...
+  sigset_t set, old;
+  __sigemptyset(&set);
+  __sigaddset(&set, SIGCHLD);
+  assert(sigprocmask(SIG_BLOCK, &set, &old) == 0);
+
+  for (int i = 0; i < NCHLD; i++) {
+    pid_t pid = fork();
+    assert(pid >= 0);
+    if (!pid) {
+      sleep(i & 0x1);
```
Comment (cahirwpz) [inline bin/utest/sigaction.c:87]: Don't use sleep in tests, unless there's really no other choice. I don't think you absolutely need it here. We prefer `sched_yield` in similar situations.

File: bin/utest/sigaction.c [hunk 10]
```diff
@@ -45,3 +53,47 @@ int test_sigaction_handler_returns(void) {
 
   return 0;
 }
+
+#define NCHLD 7
```
Comment (cahirwpz) [inline bin/utest/sigaction.c:57]: Why do you need to start several subprocesses to test `siginfo` functionality?

File: include/sys/siginfo.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+#define ksi_signo ksi_info.si_signo
+#define ksi_code ksi_info.si_code
+#define ksi_errno ksi_info.si_errno
+
+#define ksi_pid ksi_info.si_pid
+#define ksi_uid ksi_info.si_uid
+
+#define ksi_addr ksi_info.si_addr
+#define ksi_trap ksi_info.si_trap
+#define ksi_trap2 ksi_info.si_trap2
```
Comment (cahirwpz) [inline include/sys/siginfo.h:83]: `ksi_trap2` and `ksi_trap3` are not used in our code. Please remove them.

File: include/sys/signal.h [hunk 12]
```diff
@@ -73,6 +74,13 @@ typedef struct pgrp pgrp_t;
 typedef struct thread thread_t;
 typedef struct ctx ctx_t;
 
+/*! \brief Notify the parent of a change in the child's status.
```
Comment (cahirwpz) [inline include/sys/signal.h:77]: We have another function `proc_wakeup_parent` that has very similar description. This one sends a signal and wakes up the parent.

**Review by cahirwpz — COMMENTED** [review]
I have some remarks mainly concerning user-space tests.

**Review by cahirwpz — APPROVED** [review]
LGTM I'm waiting for tests in another PR.

**Pull request merged**